espnet/nets/pytorch_backend/k2_ctc.py

```diff
@@ -0,0 +1,168 @@
+        (tot_score, tot_frames, all_frames) = get_tot_objf_and_num_frames(
+            tot_scores, supervision_segments[:, 2]
+        )
+        return -tot_score
+
+
+def build_ctc_topo(tokens: List[int]) -> k2.Fsa:
+    """Build CTC topology.
+    A token which appears once on the right side (i.e. olabels) may
+    appear multiple times on the left side (ilabels), possibly with
+    epsilons in between.
+    When 0 appears on the left side, it represents the blank symbol;
+    when it appears on the right side, it indicates an epsilon. That
+    is, 0 has two meanings here.
+    Args:
+      tokens:
+        A list of tokens, e.g., phones, characters, etc.
+    Returns:
+      Returns an FST that converts repeated tokens to a single token.
+    """
+    assert 0 in tokens, "We assume 0 is ID of the blank symbol"
+
+    num_states = len(tokens)
+    final_state = num_states
+    arcs = ""
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                arcs += f"{i} {i} {tokens[i]} 0 0.0\n"
+            else:
+                arcs += f"{i} {j} {tokens[j]} {tokens[j]} 0.0\n"
+        arcs += f"{i} {final_state} -1 -1 0.0\n"
+    arcs += f"{final_state}"
+    ans = k2.Fsa.from_str(arcs)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+class CtcTrainingGraphCompiler(object):
+    def __init__(self, odim: int, device: torch.device):
+        """
+        Args:
+        odim:
+          Output dimension of CTC linear layer, len(symbol_list) + 2 (<blank> and <eos>).
+        """
+
+        self.dim = odim
+        self.device = device
+        self.ctc_topo_inv = k2.arc_sort(
+            build_ctc_topo(list(range(self.dim))).invert_().to(self.device)
+        )
+
+    def compile(self, texts: torch.Tensor, texts_lengths: torch.Tensor) -> k2.Fsa:
+        texts_lengths = torch.cat([torch.tensor([0]), texts_lengths])
+        texts_end_index = torch.cumsum(texts_lengths, 0)
+
+        decoding_graphs = k2.create_fsa_vec(
+            [
+                self.compile_one_and_cache(
+                    texts[texts_end_index[i] : texts_end_index[i + 1]]
+                )
+                for i in range(texts_lengths.shape[0] - 1)
+            ]
+        )
+
+        # make sure the gradient is not accumulated
+        decoding_graphs.requires_grad_(False)
+        return decoding_graphs
+
+    @lru_cache(maxsize=100000)
+    def compile_one_and_cache(self, text: torch.Tensor) -> k2.Fsa:
+        label_graph = k2.linear_fsa(text.tolist(), self.device)
+        label_graph = k2.add_epsilon_self_loops(label_graph)
+        decoding_graph = k2.intersect(
+            self.ctc_topo_inv, label_graph, treat_epsilons_specially=False
+        )
+        decoding_graph = k2.connect(decoding_graph.invert_())
+        return decoding_graph
+
+
+def get_tot_objf_and_num_frames(
+    tot_scores: torch.Tensor, frames_per_seq: torch.Tensor
+) -> Tuple[float, int, int]:
```

Comment:
The type hints for the return values are not correct. The three values are tensors, not scalars.